storage/redis/peer_store.go

```diff
@@ -0,0 +1,773 @@
+			}
+
+			peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+			numWant--
+		}
+
+		// Append leechers until we reach numWant.
+		if numWant > 0 {
+			announcerPK := newPeerKey(announcer)
+			for _, pk := range conLeechers {
+				if pk == announcerPK {
+					continue
+				}
+
+				if numWant == 0 {
+					break
+				}
+
+				peers = append(peers, decodePeerKey(serializedPeer(pk.([]byte))))
+				numWant--
+			}
+		}
+	}
+
+	return
+}
+
+func (ps *peerStore) ScrapeSwarm(ih bittorrent.InfoHash, af bittorrent.AddressFamily) (resp bittorrent.Scrape) {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped redis store")
+	default:
+	}
+
+	resp.InfoHash = ih
+	addressFamily := af.String()
+	encodedInfoHash := ih.String()
+	encodedLeecherInfoHash := ps.leecherInfohashKey(addressFamily, encodedInfoHash)
+	encodedSeederInfoHash := ps.seederInfohashKey(addressFamily, encodedInfoHash)
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	leechersLen, err := conn.Do("HLEN", encodedLeecherInfoHash)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encodedLeecherInfoHash,
+			"error": err,
+		})
+		return
+	}
+
+	seedersLen, err := conn.Do("HLEN", encodedSeederInfoHash)
+	if err != nil {
+		log.Error("storage: Redis HLEN failure", log.Fields{
+			"Hkey":  encodedSeederInfoHash,
+			"error": err,
+		})
+		return
+	}
+
+	resp.Incomplete = uint32(leechersLen.(int64))
+	resp.Complete = uint32(seedersLen.(int64))
+
+	return
+}
+
+// collectGarbage deletes all Peers from the PeerStore which are older than the
+// cutoff time.
+//
+// This function must be able to execute while other methods on this interface
+// are being executed in parallel.
+func (ps *peerStore) collectGarbage(cutoff time.Time) error {
+	select {
+	case <-ps.closed:
+		return nil
+	default:
+	}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	cutoffUnix := cutoff.UnixNano()
+	start := time.Now()
+
+	for _, group := range ps.groups() {
+		// list all infohashes in the group
+		infohashesList, err := conn.Do("HKEYS", group)
+		if err != nil {
+			return err
+		}
+		infohashes := infohashesList.([]interface{})
+
+		for _, ih := range infohashes {
+			ihStr := string(ih.([]byte))
+			isSeeder := len(ihStr) > 5 && ihStr[5:6] == "S"
+
+			// list all (peer, timeout) pairs for the ih
+			ihList, err := conn.Do("HGETALL", ihStr)
+			if err != nil {
+				return err
+			}
+			conIhList := ihList.([]interface{})
+
+			var pk serializedPeer
+			for index, ihField := range conIhList {
+				if index%2 == 1 { // value
+					mtime, err := strconv.ParseInt(string(ihField.([]byte)), 10, 64)
+					if err != nil {
+						return err
+					}
+					if mtime <= cutoffUnix {
+						ret, err := redis.Int64(conn.Do("HDEL", ihStr, pk))
+						if err != nil {
+							return err
+						}
+						// Maybe this peer has been removed since we got the list above with `HGETALL`.
+						if ret == 0 {
+							continue
+						}
+						// DECR seeder/leecher counter
```

Comment:
Performance optimization: keep track of how many peers you removed, then only decrement once at the end of the infohash loop.